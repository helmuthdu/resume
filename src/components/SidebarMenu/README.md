# SidebarMenu component

Sidebar menu component.

## Source

    <SidebarMenu visible={boolean}  />
