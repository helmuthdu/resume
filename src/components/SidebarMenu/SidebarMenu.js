// @flow
import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { Dispatch } from 'redux';
import { MenuItem, MenuProps } from '../Menu';
import { UIStoreState } from '../../store/modules/ui';
import { Icon, Menu, Sidebar } from 'semantic-ui-react';

interface SidebarMenuProps extends MenuProps {
  visible?: boolean;
  dispatch?: Dispatch<any>;
  Link: any;
}

export const SidebarMenu = ({ items, pathname, visible }: SidebarMenuProps) => {
  const isActive = (item: MenuItem) => (item.exact) ? pathname === item.path : pathname.startsWith(item.path);
  const activeItem = items.find((item: MenuItem) => isActive(item)) || {};
  return (
    <Sidebar as={Menu} animation="slide along" width="thin" visible={visible} icon="labeled" vertical inverted={activeItem.inverted}>
      {items.map((item) => {
        const active = isActive(item);
        return (
          <Menu.Item as={Link} to={item.path} active={active} key={item.path}>
            <Icon name={item.icon}/>
            {item.name}
          </Menu.Item>
        );
      })}
    </Sidebar>
  );
};

const mapStateToProps = (state: UIStoreState) => ({
  visible: state.isSidebarVisible
});

export default connect(mapStateToProps)(SidebarMenu);
