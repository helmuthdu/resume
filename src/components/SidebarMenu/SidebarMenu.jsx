// @flow
import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { bindActionCreators, Dispatch } from 'redux';
import { Icon, Menu, Sidebar } from 'semantic-ui-react';
import { toggleSidebar } from '../../store/modules/ui';
import { MenuItem, MenuProps } from '../Menu/Menu';

interface SidebarMenuProps extends MenuProps {
  items: MenuItem[];
  pathname: any;
  toggleSidebar: Dispatch;
  visible?: boolean;
}

export const SidebarMenu = ({ items, pathname, visible, toggleSidebar }: SidebarMenuProps) => {
  const isActive = (item: MenuItem) => (item.exact ? pathname === item.path : pathname.startsWith(item.path));
  const activeItem = items.find((item: MenuItem) => isActive(item)) || {};
  return (
    <Sidebar as={Menu} visible={visible} vertical animation="uncover" inverted={activeItem.inverted}>
      {items.map(item => {
        const active = isActive(item);
        return (
          <Menu.Item as={Link} to={item.path} active={active} key={item.path} onClick={toggleSidebar}>
            <Icon name={item.icon} />
            {item.name}
          </Menu.Item>
        );
      })}
    </Sidebar>
  );
};

const mapStateToProps = (state: any) => ({
  visible: state.ui.isSidebarVisible
});

const mapDispatchToProps = (dispatch: Dispatch) =>
  bindActionCreators(
    {
      toggleSidebar
    },
    dispatch
  );

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(SidebarMenu);
