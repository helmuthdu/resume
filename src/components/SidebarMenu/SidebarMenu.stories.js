import { action } from '@storybook/addon-actions';
import { text, withKnobs } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react';
import React from 'react';
import SidebarMenu from './SidebarMenu';

const withReadme = require('storybook-readme/with-readme').default;
const SidebarMenuReadme = require('./README.md');

const items = [
  { name: 'Home', path: '/', exact: true, icon: 'home' },
  { name: 'About', path: '/about/', exact: true, icon: 'info circle' },
  { name: 'Blog', path: '/blog/', exact: false, icon: 'newspaper' }
];

const LinkStub = props => (
  <div {...props} onClick={action(props.to.toString())}>
    {props.children}
  </div>
);

storiesOf('SidebarMenu', module)
  .addDecorator(withKnobs)
  .addDecorator(withReadme(SidebarMenuReadme))
  .add('default', () => {
    const pathname = text('pathname', '/');

    return <SidebarMenu Link={LinkStub} items={items} pathname={pathname} visible />;
  });
