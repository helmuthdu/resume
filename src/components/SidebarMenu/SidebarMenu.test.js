import { configure, render } from 'enzyme';
import 'jest';
import React from 'react';
import { withRouter } from 'react-router';
import { SidebarMenu } from './SidebarMenu';

// Configure enzyme with react 16 adapter
const Adapter: any = require('enzyme-adapter-react-16');
configure({ adapter: new Adapter() });

const items = [
  { name: 'Home', path: '/', exact: true },
  { name: 'About', path: '/about/', exact: true },
  { name: 'Blog', path: '/blog/', exact: false }
];

describe('SidebarMenu component', () => {
  it('should render correctly', () => {

    let visible = true;
    const toggleSidebar = () => visible = !visible;
    const wrapper = render(withRouter(<SidebarMenu pathname="/" items={items} visible={visible} toggleSidebar={toggleSidebar}/>));
    expect(wrapper).toMatchSnapshot();
  });
});
