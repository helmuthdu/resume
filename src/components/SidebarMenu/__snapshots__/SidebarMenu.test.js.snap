// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SidebarMenu component should render correctly 1`] = `
initialize {
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;

exports[`SidebarMenu component should render correctly 2`] = `
initialize {
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;

exports[`SidebarMenu component should t 1`] = `
initialize {
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;

exports[`SidebarMenu component should toggle  1`] = `
initialize {
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;

exports[`SidebarMenu component should toggle 1`] = `
initialize {
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;

exports[`SidebarMenu component should toggle side 1`] = `
initialize {
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;

exports[`SidebarMenu component should toggle sidebar 1`] = `
initialize {
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
